File: webclient/js/Orientation.js
// Orientation.js

var TAGGER = TAGGER || {}

TAGGER.orientation = (function () {
   var o = {};

   // converts cartesian to spherical
   function sCoords (cCoords) {
      var sCoords = {};
      var flat_r = Math.sqrt(cCoords.x * cCoords.x + cCoords.z * cCoords.z);
      sCoords.theta = Math.atan2(-cCoords.z, cCoords.x);
      sCoords.phi   = Math.atan2(cCoords.y, flat_r);
      return sCoords;
   }
   o.sCoords = sCoords;

   // converts spherical to cartesian
   function cCoords (sCoords) {
      var cCoords = {};
      var r = sCoords.r || 1;
      var flat_r = r * Math.cos(sCoords.phi);
      cCoords.y  = r * Math.sin(sCoords.phi);
      cCoords.x  =  flat_r * Math.cos(sCoords.theta);
      cCoords.z  = -flat_r * Math.sin(sCoords.theta);
      return cCoords;
   }
   o.cCoords = cCoords;

   // rotates a cartesian vector about the z axis
   function rotz (p, theta) {
      var cost = Math.cos(theta);
      var sint = Math.sin(theta);
      return { x: p.x * cost - p.y * sint,
               y: p.y * cost + p.x * sint,
               z: p.z };
   }


   // Field of view and derivative variables.
   o.fov_theta = 0.0;
   o.fov_phi   = 0.0;
   o.spos  = { x: 0, y: 0, z: 0 };
   o.right = { x: 0, y: 0, z: 0 };
   o.down  = { x: 0, y: 0, z: 0 };

   // the direction we're facing (in the ground plane)
   o.view        = { theta: 0.0, phi: 0.0 };

   // Used for dragging. Set by cacheView.
   o.cachedView  = { theta: 0.0, phi: 0.0 };
   var cachedSpos  = { x: 0, y: 0, z: 0 };
   var cachedRight = { x: 0, y: 0, z: 0 };
   var cachedDown  = { x: 0, y: 0, z: 0 };


   // sets the field of view
   o.setFov = function (params) {
      var fovt = params.theta;
      var fovp = params.phi;
      o.fov_theta = fovt;
      o.fov_phi = fovp;
      var tant = Math.tan(fovt / 2);
      var tanp = Math.tan(fovp / 2);
      o.spos  = { x: 1, y:    tanp, z:   -tant };
      o.right = { x: 0, y:       0, z:  2*tant };
      o.down  = { x: 0, y: -2*tanp, z:       0 };
   };

   // caches the current view
   o.cacheView = function () {
      o.cachedView = { theta: o.view.theta, phi: o.view.phi };
      cachedSpos = rotz(o.spos, o.cachedView.phi);
      // we don't need to rotate this because it lies along the z axis
      cachedRight = o.right;
      cachedDown = rotz(o.down, o.cachedView.phi);
   };

   // retrieves cached view
   o.getCachedView = function () {
      return o.cachedView;
   };

   // changes view to look at given coords
   function lookAt (sCoords) {
      o.view = sCoords;
   }
   o.lookAt = lookAt;

   // returns the current view in cartesian coordinates
   o.cView = function () {
      return cCoords(o.view);
   };

   // Given normalized screen coordinates, this function
   // returns the sperhical coordinates of the mouse's projection.
   function project (screenCoords) {
      var mcoords = { x: cachedSpos.x + screenCoords.y * cachedDown.x,
                      y: cachedSpos.y + screenCoords.y * cachedDown.y,
                      z: cachedSpos.z + screenCoords.x * cachedRight.z };
      return sCoords(mcoords);
   }
   o.project = project;

   return o;
}());


TAGGER.mouse = (function () {
   var mouse = {};

   // mouse variables
   down_theta  = 0.0;
   down_phi    = 0.0;
   track_theta = 0.0;
   track_phi   = 0.0;

   mouse.downtheta = function () { return down_theta; };
   mouse.downphi = function () { return down_phi; };
   mouse.ttheta = function () { return track_theta; };
   mouse.tphi = function () { return track_phi; };

   mouse.down = function (loc) {
      down_theta = loc.theta;
      down_phi   = loc.phi;
   };

   function track (loc) {
      track_theta = loc.theta;
      track_phi   = loc.phi;
   }
   mouse.track = track;

   mouse.delta = function () {
      return {
         theta: track_theta - down_theta,
         phi:   track_phi   - down_phi
      };
   };

   mouse.box = function () {
      return TAGGER.common.box({
         theta1: down_theta,
         phi1: down_phi,
         theta2: track_theta,
         phi2: track_phi
      });
   };

   return mouse;
}());

   

File: legacy/python/Orientation.py
#!/usr/bin/python

from common import *

class Vector:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __add__(self, a):
        return Vector(self.x + a.x, self.y + a.y, self.z + a.z)

    def __sub__(self, a):
        return Vector(self.x - a.x, self.y - a.y, self.z - a.z)
        
    def __rmul__(self, s):
        return Vector(s*self.x, s*self.y, s*self.z)

# Rotates a Vector about the x axis
def rotx(theta, p):
    cost = cos(theta)
    sint = sin(theta)
    return Vector(p.x, p.y*cost - p.z*sint, p.z*cost + p.y*sint)

# Rotates a Vector about the y axis
def roty(theta, p):
    cost = cos(theta)
    sint = sin(theta)
    return Vector(p.x*cost + p.z*sint, p.y, p.z*cost - p.x*sint)

# Rotates a Vector about the z axis
def rotz(theta, p):
    cost = cos(theta)
    sint = sin(theta)
    return Vector(p.x*cost - p.y*sint, p.y*cost + p.x*sint, p.z)

# Converts Cartesian to Spherical
def cartesianToSpherical(point):
    r     = math.sqrt(point.x*point.x + point.y*point.y + point.z*point.z)
    theta = atan2(point.y, point.x)
    phi   = asin(point.z / r)
    return SPoint(theta, phi, r)

# Converts Spherical to Cartesian
def sphericalToCartesian(point):
    r = point.r * cos(point.phi) # distance from origin of point projected onto xy plane
    x = r * cos(point.theta)
    y = r * sin(point.theta)
    z = point.r * sin(point.phi)
    return CPoint(x, y, z)

# All orientation related variables needed by the SVApp
class Orientation:
    def __init__(self):
        # direction we're facing
        self.theta      = 0.0   # degrees counterclockwise from east, along the plane of the ground
        self.phi        = 0.0   # degrees up or down from the ground

        # orientation of panorama relative to cardinal directions
        # these are just copies of the same variables in SVApp.pano
        self.pano_yaw   = 0.0   # yaw of the center of the current panorama
        self.tilt_yaw   = 0.0   # yaw of the direction of greatest slope (of the ground)
        self.tilt_pitch = 0.0   # pitch of the ground relative to cardinal plane when facing tilt_yaw

        # field of view
        self.fov_theta  = 0.0   # degrees visible along the azimuth
        self.fov_phi    = 0.0   # degrees visible up and down
        self.fov_mult   = 0.9   # fov is multiplied or divided by this when zooming in / out
        self.fov_min    = 7.5   # mimimum allowed fov
        self.fov_max    = 140.  # maximum allowed fov
        self.spos       = Vector(0,0,0)
        self.sright     = Vector(0,0,0)
        self.sleft      = Vector(0,0,0)

    # Makes convenience copies of the panorama's orientation
    def setPano(self, pano):
        self.pano_yaw = pano.panoYaw
        self.tilt_yaw = pano.tiltYaw
        self.tilt_pitch = pano.tiltPitch

    # Defines the field of view. Necessary for figuring out spherical mouse coordinates.
    def setFov(self, fovt, fovp):
        self.fov_theta = fovt
        self.fov_phi   = fovp
        tant = tan(self.fov_theta / 2.)
        tanp = tan(self.fov_phi / 2.)
        x = 1.0 / math.sqrt( 1 + tant*tant + tanp*tanp )
        self.spos   = Vector(x, x*tant, x*tanp)
        self.sright = Vector(0, -2*x*tant, 0)
        self.sdown  = Vector(0, 0, -2*x*tanp)

    def zoomIn(self):
        fovt = self.fov_theta * self.fov_mult
        fovp = self.fov_phi * self.fov_mult
        if fovt < self.fov_min: fovt = self.fov_min
        if fovp < self.fov_min: fovp = self.fov_min
        self.setFov(fovt, fovp)

    def zoomOut(self):
        fovt = self.fov_theta / self.fov_mult
        fovp = self.fov_phi / self.fov_mult
        if fovt > self.fov_max: fovt = self.fov_max
        if fovp > self.fov_max: fovp = self.fov_max
        self.setFov(fovt, fovp)

    # Takes the pixel coordinates of the mouse and calculates where on the sphere it points.
    # mousex and mousey should both be normalized to the range (0, 1),
    # and measured from the top left of the window (x horizontal, y vertical).
    # The returned angles are measured in the ground plane: counterclockwise from east and
    # up from the ground.
    def screenCoords(self, right, down):
        m = cartesianToSpherical( rotz(self.theta, roty(-self.phi, self.spos + right*self.sright + down*self.sdown)) )
        # Right now m.theta is in the range (-180, 180]. We want to put it in the range
        # (self.phi - 180, self.phi + 180] so the branch point is as far away as it can be.
        theta = m.theta
        if theta < self.theta - 180.: theta += 360.
        if theta > self.theta + 180.: theta -= 360.
        return SDirection(theta, m.phi)

    def cartesianScreenCoords(self):
        return rotz(self.theta, roty(-self.phi, self.spos))

    # Finds the angle in the ground plane from east to the tilt axis
    def tiltedYaw(self):
        if self.tilt_yaw <= 180.0:
            bigyaw = False
            cosy = cos(self.tilt_yaw)
        else:
            bigyaw = True
            cosy = cos(self.tilt_yaw - 180.0)
        cosp = cos(self.tilt_pitch)
        sinp = sin(self.tilt_pitch)
        temp = acos( cosy / math.sqrt(cosp*cosp + cosy*cosy * sinp*sinp) )
        if bigyaw:
            return temp + 180.0
        else:
            return temp

    # Finds the pitch angle that places the x axis in the ground plane
    # when the x axis is theta degrees councterclockwise from east in the cardinal plane.
    def tiltedPitch(self, theta):
        cosp = cos(self.tilt_pitch)
        sinp = sin(self.tilt_pitch)
        siny = sin(theta)
        temp = acos( cosp / math.sqrt(cosp*cosp + sinp*sinp * siny*siny) )
        if theta > 180.0:
            return -temp
        else:
            return temp

